generator/src/main/java/com/graphicsfuzz/generator/fuzzer/OpaqueExpressionGenerator.java

```diff
@@ -691,6 +697,105 @@ public Expr apply(Expr expr, BasicType type, boolean constContext, int depth,
 
   }
 
+  /**
+   * Identity transformation for integer types (both unsigned and signed, and their vectors) that
+   * double bitwise inverts an integer, producing the same integer as output. When performed,
+   * transforms an expression, e, such that:
+   *    e -> ~(~(expr)).
+   */
+  private class IdentityBitwiseNotNot extends AbstractIdentityTransformation {
+    private IdentityBitwiseNotNot() {
+      super(BasicType.allIntegerTypes(), false);
+    }
+
+    @Override
+    public Expr apply(Expr expr, BasicType type, boolean constContext, int depth,
+                      Fuzzer fuzzer) {
+      assert BasicType.allIntegerTypes().contains(type);
+      // Invert once
+      Expr result = new UnaryExpr(new ParenExpr(applyIdentityFunction(expr, type,
+          constContext,
+          depth, fuzzer)), UnOp.BNEG);
+      // Invert again
+      result = new UnaryExpr(new ParenExpr(applyIdentityFunction(result, type, constContext,
+          depth, fuzzer)), UnOp.BNEG);
+      return identityConstructor(expr, result);
+    }
+  }
+
+  /**
+   * Identity transformation for integer types (both unsigned and signed, and their vectors) that
+   * ORs an integer with itself, producing the same integer as output. When performed, transforms an
+   * expression, e, such that:
+   *    e -> (e) | (e).
+   */
+  private class IdentityBitwiseOr extends AbstractIdentityTransformation {
+    private IdentityBitwiseOr() {
+      super(BasicType.allIntegerTypes(), false);
+    }
+
+    @Override
+    public Expr apply(Expr expr, BasicType type, boolean constContext, int depth,
+                      Fuzzer fuzzer) {
+      assert BasicType.allIntegerTypes().contains(type);
```

Comment:
Thanks for the example in the comment, but can you make it more explicit as to what is being transformed, e.g. "for example transforming ... to ... | ... would not preserve semantics."

(The use of the assignment operator in the example looks a bit odd to me - this is the kind of thing that would not be side-effect free and that therefore we shouldn't do, as d would get assigned to twice not once.)

Reply:
Sorry for the ambiguity - the reason I put that as a specific example is because that is the exact expression I ran into issues with when running tests for that identity. The above fix, to ensure the expression must be side effect free, should fix this issue. I didn't really think about the fact that the assignment would have side effects, so I just patched that case instead of just preventing side effects for all expressions. Whoops.

Reply:
Additionally, upon retesting without parentheses, I can't see from the expressions what exactly is going wrong with the failing tests. The one thing they have in common is that they're dealing with bitwise OR in ternary expressions.

I've cut down the comment (since I can't even find a good concrete example) to give the basic idea, and I was able to move the parentheses to outside the OR expression (instead of surrounding the two mutated expressions) without issue.